File: src/web/src/lib/types/consultation.ts
/**
 * @fileoverview Consultation type definitions for AUSTA SuperApp virtual care platform
 * @version 1.0.0
 * @license HIPAA-compliant
 */

import { Room, LocalTrack, RemoteParticipant } from 'twilio-video'; // v2.27.0
import { IUser, UserRole } from '../types/user';

/**
 * Enum defining available consultation modalities
 * Supports multiple communication channels for virtual care
 */
export enum ConsultationType {
    VIDEO = 'VIDEO',
    AUDIO = 'AUDIO',
    CHAT = 'CHAT'
}

/**
 * Enum defining possible consultation session states
 * Tracks complete lifecycle of virtual care sessions
 */
export enum ConsultationStatus {
    SCHEDULED = 'SCHEDULED',
    WAITING = 'WAITING',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETED = 'COMPLETED',
    CANCELLED = 'CANCELLED',
    FAILED = 'FAILED'
}

/**
 * Enum defining WebRTC connection quality levels
 * Used for monitoring and reporting session quality
 */
export enum ConnectionQuality {
    EXCELLENT = 'EXCELLENT',
    GOOD = 'GOOD',
    FAIR = 'FAIR',
    POOR = 'POOR',
    DISCONNECTED = 'DISCONNECTED'
}

/**
 * Interface for tracking consultation participant details
 * Includes comprehensive monitoring of participant state and metrics
 */
export interface IConsultationParticipant {
    userId: string;
    role: UserRole;
    joinedAt: Date;
    leftAt: Date | null;
    isAudioEnabled: boolean;
    isVideoEnabled: boolean;
    connectionQuality: ConnectionQuality;
    deviceInfo: Record<string, string>;
    networkStats: Record<string, number>;
}

/**
 * Interface for virtual care consultation data
 * Implements HIPAA-compliant session management with audit tracking
 */
export interface IConsultation {
    id: string;
    type: ConsultationType;
    patientId: string;
    providerId: string;
    scheduledStartTime: Date;
    actualStartTime: Date | null;
    endTime: Date | null;
    status: ConsultationStatus;
    participants: readonly IConsultationParticipant[];
    healthRecordId: string | null;
    roomSid: string | null;
    metadata: Record<string, unknown>;
    securityMetadata: Record<string, string>;
    auditLog: readonly Record<string, any>[];
}

/**
 * Interface for WebRTC room management
 * Implements secure video/audio streaming with encryption verification
 */
export interface IConsultationRoom {
    room: Room;
    localTracks: readonly LocalTrack[];
    participants: ReadonlyMap<string, RemoteParticipant>;
    connectionState: ConnectionQuality;
    encryptionEnabled: boolean;
}

/**
 * Type guard to check if a participant is a healthcare provider
 * @param participant The consultation participant to check
 * @returns boolean indicating if participant is a provider
 */
export const isProvider = (participant: IConsultationParticipant): boolean => {
    return participant.role === UserRole.PROVIDER;
};

/**
 * Type guard to check if a consultation is active
 * @param consultation The consultation to check
 * @returns boolean indicating if consultation is currently active
 */
export const isActiveConsultation = (consultation: IConsultation): boolean => {
    return consultation.status === ConsultationStatus.IN_PROGRESS ||
           consultation.status === ConsultationStatus.WAITING;
};

/**
 * Type guard to check if a consultation room is secure
 * @param room The consultation room to check
 * @returns boolean indicating if room meets security requirements
 */
export const isSecureRoom = (room: IConsultationRoom): boolean => {
    return room.encryptionEnabled && room.connectionState !== ConnectionQuality.DISCONNECTED;
};

File: src/android/app/src/main/kotlin/com/austa/superapp/features/virtualcare/domain/models/Consultation.kt
package com.austa.superapp.features.virtualcare.domain.models

import com.austa.superapp.features.auth.domain.models.User
import com.austa.superapp.core.utils.ValidationUtils
import com.austa.superapp.core.constants.AppConstants
import com.austa.security.SecurityUtils // v1.0.0
import kotlinx.serialization.Serializable // v1.5.1
import java.util.Date
import java.util.UUID

/**
 * HIPAA-compliant data class representing a virtual care consultation session.
 * Implements secure video consultation with comprehensive audit trails and role-based access control.
 */
@Serializable
data class Consultation(
    val id: String = UUID.randomUUID().toString(),
    val patientId: String,
    val providerId: String,
    val scheduledStartTime: Date,
    val actualStartTime: Date? = null,
    val endTime: Date? = null,
    val status: ConsultationStatus = ConsultationStatus.SCHEDULED,
    val participants: MutableList<ConsultationParticipant> = mutableListOf(),
    val healthRecordId: String? = null,
    val twilioRoomSid: String? = null,
    val metadata: ConsultationMetadata = ConsultationMetadata(),
    val auditTrail: MutableList<AuditEntry> = mutableListOf(),
    private val sensitiveInfo: EncryptedData? = null
) {
    init {
        require(ValidationUtils.validateHealthData(mapOf(
            "patientId" to patientId,
            "providerId" to providerId
        )).isValid) { "Invalid participant IDs" }

        // Add initial audit entry
        auditTrail.add(AuditEntry(
            action = "CONSULTATION_CREATED",
            timestamp = Date(),
            performedBy = "SYSTEM",
            details = "Consultation scheduled for $scheduledStartTime"
        ))
    }

    /**
     * Consultation status states with HIPAA compliance tracking
     */
    enum class ConsultationStatus {
        SCHEDULED,
        IN_PROGRESS,
        COMPLETED,
        CANCELLED,
        NO_SHOW,
        TECHNICAL_ISSUES
    }

    /**
     * HIPAA-compliant participant information
     */
    @Serializable
    data class ConsultationParticipant(
        val userId: String,
        val role: User.UserRole,
        val joinTime: Date? = null,
        val leaveTime: Date? = null,
        val connectionQuality: ConnectionQuality = ConnectionQuality.UNKNOWN,
        val deviceInfo: DeviceInfo
    )

    /**
     * Connection quality monitoring for performance optimization
     */
    enum class ConnectionQuality {
        EXCELLENT,
        GOOD,
        FAIR,
        POOR,
        DISCONNECTED,
        UNKNOWN
    }

    /**
     * Device information for audit and troubleshooting
     */
    @Serializable
    data class DeviceInfo(
        val deviceId: String,
        val platform: String,
        val osVersion: String,
        val appVersion: String,
        val networkType: String
    )

    /**
     * Metadata for consultation session
     */
    @Serializable
    data class ConsultationMetadata(
        val specialtyType: String? = null,
        val consultationType: String = "GENERAL",
        val priority: Priority = Priority.NORMAL,
        val notes: String? = null,
        val recordingEnabled: Boolean = false,
        val maxDuration: Int = AppConstants.VIRTUAL_CARE.SESSION_TIMEOUT_MINUTES,
        val technicalConfig: TechnicalConfig = TechnicalConfig()
    )

    /**
     * Technical configuration for video consultation
     */
    @Serializable
    data class TechnicalConfig(
        val videoQuality: String = AppConstants.VIRTUAL_CARE.MAX_VIDEO_QUALITY,
        val audioBitrate: Int = AppConstants.VIRTUAL_CARE.MAX_AUDIO_BITRATE,
        val videoBitrate: Int = AppConstants.VIRTUAL_CARE.MAX_VIDEO_BITRATE,
        val frameRate: Int = AppConstants.VIRTUAL_CARE.FRAME_RATE,
        val echoCancellation: Boolean = AppConstants.VIRTUAL_CARE.ECHO_CANCELLATION,
        val noiseSuppression: Boolean = AppConstants.VIRTUAL_CARE.NOISE_SUPPRESSION
    )

    /**
     * Priority levels for consultation scheduling
     */
    enum class Priority {
        URGENT,
        HIGH,
        NORMAL,
        LOW
    }

    /**
     * Audit trail entry for HIPAA compliance
     */
    @Serializable
    data class AuditEntry(
        val timestamp: Date,
        val action: String,
        val performedBy: String,
        val details: String,
        val ipAddress: String = "INTERNAL",
        val success: Boolean = true
    )

    /**
     * Encrypted sensitive data container
     */
    @Serializable
    data class EncryptedData(
        val encryptedContent: String,
        val iv: String,
        val algorithm: String = AppConstants.SECURITY.ENCRYPTION_ALGORITHM
    )

    /**
     * Checks if consultation is currently active
     * @return Boolean indicating if consultation is in progress
     */
    fun isActive(): Boolean {
        return status == ConsultationStatus.IN_PROGRESS &&
                actualStartTime != null &&
                endTime == null &&
                participants.any { it.leaveTime == null }
    }

    /**
     * Calculates consultation duration with caching optimization
     * @return Long duration in minutes, or null if consultation hasn't ended
     */
    fun getDuration(): Long? {
        if (actualStartTime == null || endTime == null) return null
        return (endTime.time - actualStartTime.time) / (60 * 1000)
    }

    /**
     * Validates if user has permission to access this consultation
     * @param userId String ID of user attempting access
     * @param role User.UserRole role of user attempting access
     * @return Boolean indicating if access is permitted
     */
    fun canAccess(userId: String, role: User.UserRole): Boolean {
        return when (role) {
            User.UserRole.PROVIDER -> userId == providerId
            User.UserRole.PATIENT -> userId == patientId
            User.UserRole.ADMIN -> true
            else -> false
        }
    }
}